### PR TITLE
feat: token paymaster integration

### PR DESCRIPTION
**Pull request by VGabriel45**
## PR-Codex overview
This PR focuses on enhancing the Biconomy Token Paymaster functionality within the SDK, including updates to the paymaster context, the introduction of new methods for token management, and improvements to user operation handling.

### Detailed summary
- Updated `BICONOMY_TOKEN_PAYMASTER` address in `Constants.ts`.
- Renamed `biconomyPaymasterContext` to `biconomySponsoredPaymasterContext`.
- Added `getAllowance` function in `Utils.ts`.
- Introduced `getSupportedTokens` and `getTokenPaymasterQuotes` methods for token management.
- Updated `NexusClient` to use `PaymasterContext`.
- Enhanced user operation preparation and sending with token paymaster functionality.
- Added tests for token paymaster operations in `createBicoPaymasterClient.test.ts`.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/sdk/clients/createBicoBundlerClient.ts
Comment (joepegler) [inline src/sdk/clients/createBicoBundlerClient.ts:24]: Why is this paymaster context in the createBundlerClient file?

File: src/test/playground.test.ts
Comment (joepegler) [inline src/test/playground.test.ts:83]: Why remove the old test? If this test is important can you keep it in cicd? The playground doesn't run in cicd

**Comment by joepegler** [comment]
Please write a token paymaster 'dx' test so I can see what the devEx is like.

File: src/sdk/clients/createNexusClient.ts
Comment (joepegler) [inline src/sdk/clients/createNexusClient.ts:222]: Why is there a hardcoded token address in here? Please remove it.

File: src/test/playground.test.ts
Comment (joepegler) [inline src/test/playground.test.ts:118]: Please move this out of the playground

File: src/test/playground.test.ts
Comment (joepegler) [inline src/test/playground.test.ts:157]: Why did we remove this?

File: src/test/testUtils.ts
Comment (joepegler) [inline src/test/testUtils.ts:498]: Why is this commented?

File: src/test/playground.test.ts
Comment (joepegler) [inline src/test/playground.test.ts:151]: Now that a user has the quote how can he use it?

**Review by joepegler — CHANGES_REQUESTED** [review]

File: src/sdk/clients/createBundlerClient.test.ts
Comment (joepegler) [inline src/sdk/clients/createBundlerClient.test.ts:89]: Why is this skipped?

**Comment by joepegler** [comment]
Please write tests for each of the following:

1. "should retrieve all supported token addresses from the token paymaster"
2. "should retrieve all quotes from a paymaster for a user op"
3. "should select and use the fee quote to send the tx"
4. "should send a tx using token paymaster specifying only the token address and the userOp

If point #4 isn't currently possible for example, then please go ahead and wire the getFeeCall result with the call to send the transaction under the hood

**Comment by joepegler** [comment]
What did you do about token approvals?

File: src/sdk/clients/createBicoBundlerClient.ts
Comment (VGabriel45) [reply src/sdk/clients/createBicoBundlerClient.ts:24]: It wasn't used anywhere else, I can move it in the bico paymaster file, np.

File: src/sdk/clients/createNexusClient.ts
Comment (VGabriel45) [reply src/sdk/clients/createNexusClient.ts]: Resolved

File: src/test/playground.test.ts
Comment (VGabriel45) [reply src/test/playground.test.ts]: Resolved

File: src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts
Comment (joepegler) [inline src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts:108]: Small detail but can you change this to getTokenPaymasterQuotes(client, parameters) like all the other decorators?

```ts
type GetTokenPaymasterQuotesParameters = {
  userOp: UserOperation;
  tokenList: Address[]
}
```

File: src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts
Comment (joepegler) [inline src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts:136]: This doesn't come from the paymasterContext in the account somewhere? Can't users specifiy the expiry for example?

File: src/sdk/clients/decorators/tokenPaymaster/index.ts
Comment (joepegler) [inline src/sdk/clients/decorators/tokenPaymaster/index.ts:73]: returns {@TheRealTypescriptType}
Then you can omit lines 91 - 108

**Review by joepegler — CHANGES_REQUESTED** [review]
review ii

**Comment by joepegler** [comment]
build / size report are failing

File: src/sdk/clients/createNexusClient.ts
Comment (joepegler) [inline src/sdk/clients/createNexusClient.ts:227]: Don't understand this bit. Why isn't this simply `paymasterContext`? Haven't you just broken support for all other paymasters in doing this?

**Review by joepegler — CHANGES_REQUESTED** [review]

File: src/sdk/account/utils/Utils.ts
Comment (joepegler) [inline src/sdk/account/utils/Utils.ts:412]: why casting here?

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (joepegler) [inline src/sdk/clients/createBicoPaymasterClient.test.ts:237]: What if a user wants to use sendUserOperation() instead of sendTransaction()?

Comment (joepegler) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:237]: I think writing a `prepareTokenPaymasterUserOperation()` call would be a better thing than attempting to hijack what sendTransaction does here. 

Then later you could add a `sendTokenPaymasterUserOperation()` which will pair `prepareTokenPaymasterUserOperation` with (sendTransaction() or `sendUserOperation`)

**Review by joepegler — CHANGES_REQUESTED** [review]
review ii

File: src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts
Comment (VGabriel45) [reply src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts:121]: Doesn't work, the tpm needs different types for user op like string of numbers, deepHexlify transforms every field into a hex string.

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (joepegler) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:170]: There are other places in tests where the chain is not required during sendTransaction though right? Can you investigate to make sure that this isn't because of your changes?

**Comment by joepegler** [comment]
Still have failing tests 
image

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (VGabriel45) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:160]: No it's not related to my changes, we only use "sendTransaction({calls: [tx]})" inside our SDK, we haven't used "sendTransaction(tx)" anywhere, default viem behaviour is to require chain as well.

File: src/sdk/clients/createBicoPaymasterClient.ts
Comment (VGabriel45) [reply src/sdk/clients/createBicoPaymasterClient.ts:58]: Resolved

File: src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts
Comment (VGabriel45) [reply src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts:70]: Added the link but will prefer to keep the example response, adds a bit more clarity by showing an actual response example.

File: src/sdk/clients/decorators/tokenPaymaster/index.ts
Comment (VGabriel45) [reply src/sdk/clients/decorators/tokenPaymaster/index.ts:19]: Added the link but will prefer to keep the example response, adds a bit more clarity by showing an actual response example.

File: src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts
Comment (VGabriel45) [reply src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts:108]: Resolved

File: src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts
Comment (VGabriel45) [reply src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts:136]: What is the issue with "paymasterVerificationGasLimit" ? You are pointing to this in this comment but you're mentioning something else. Users can set their own expiryDuration in the "paymasterContext", if they don't, the default values will be used.

File: src/sdk/clients/decorators/tokenPaymaster/index.ts
Comment (VGabriel45) [reply src/sdk/clients/decorators/tokenPaymaster/index.ts:85]: Added the link but will prefer to keep the example response, adds a bit more clarity by showing an actual response example.

File: src/sdk/clients/createNexusClient.ts
Comment (VGabriel45) [reply src/sdk/clients/createNexusClient.ts:227]: Basically if "paymasterContext" is not undefined here, we use the fields that are set inside of it and we add the ones that never change so that the user only have to provide the fee token address and some other optional fields like expiry. If there is not paymaster context it defaults to undefined and viem will default it for sponsored mode, it does not break any support for other paymasters. 

Anyway, thanks for pointing this out, I have changed it to a more elegant approach.

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (VGabriel45) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:103]: Can you explain why "getPaymasterStubData" is being removed in our SDK ?

File: src/sdk/account/utils/Utils.ts
Comment (VGabriel45) [reply src/sdk/account/utils/Utils.ts:412]: I was seeing "unknown" type coming from readContract, now it seems like it does not complain anymore. Resolved.

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (VGabriel45) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:227]: I don't think this is a good idea. We are adding too many methods, this will become confusing and hard to use. What is the issue with using sendTransaction like this ? A user uses the same method he is used to. It doesn't make sense to add separated methods to send token paymaster user ops, we don't do this for sponsored mode.

For "sendUserOperation" the user (developer) will have to make sure he also batches an approval inside the user op if the smart account does not have enough allowance. Calling the "sendUserOperation" assumes you want to built the user op yourself, for some reason, so this means you also need to make sure the user op is valid. "sendTransaction" is an abstraction on top of user operations, we should encourage everyone to use this and only use "sendUserOperation" for special cases.

Coming back to your idea, I can add an "prepareTokenPaymasterUserOperation" or we can keep the familiar DEVX and use the already existing "prepareUserOperation" with an extra boolean parameter "withTokenPaymaster", if true our SDK injects an approve tx inside the user op, if needed.

**Comment by VGabriel45** [comment]
> Still have failing tests image

These tests are failling because there is no deposit left for this token paymaster contract on base sepolia, I asked Fil to deposit some more base sepolia eth.

**Comment by joepegler** [comment]
Failing tests and conflicts. Can you mark PRs as a draft prior while tests are failing?

**Review by joepegler — CHANGES_REQUESTED** [review]
review (iii)

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (joepegler) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:225]: If you write a 'prepareTokenPaymasterUserOperation' helper then it should be available for use without changes via either sendTransaction pr sendUserOperation. In fact that would make for a good test to write:

("should use nexusClient.prepareTokenPaymasterUserOperation to prepare a user op, and then send it using nexusClient.sendUserOperation", () => {})

Comment (joepegler) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:225]: sendTransaction doesn't do much more than simply call sendUserOperation. It's only really there as a way of keeping some form of backwards compatibility for our v4 users

**Comment by joepegler** [comment]
https://sepolia.basescan.org/vmtrace?txhash=0xaa94feb29207ce281abbb715d4666b339e9513174d74908b258f017d5d8aad07&type=gethtrace2

Stuff seems to be failing because of approvals @VGabriel45: 
image

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (joepegler) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:105]: Sure. If you leave it in place then it triggers 7767 flows by default, which unnecessarily adds additional latency. This is the way it has been designed by viem. There is more context about that here: https://biconomyworkspace.slack.com/archives/D06JEAU5G6R/p1727786219358339

**Review by joepegler — CHANGES_REQUESTED** [review]

File: src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts
Comment (joepegler) [reply src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts:141]: I was referring to the object starting on line 141. In here I might have expected that the expiryDuration might come from the end user, instead of it being hardcoded. 

Why is expiryDuration hardcoded in here? Can't you pull the paymasterContext out of the client here `client.paymasterContext` ?

File: src/sdk/clients/createNexusClient.ts
Comment (joepegler) [inline src/sdk/clients/createNexusClient.ts:129]: I don't think this works here. This is more like `BiconomyPaymasterContext`. `PaymasterContext` should be of type `unknown`, as certainly I think there's no consistency between different vendors data structures here.

File: src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts
Comment (joepegler) [inline src/sdk/clients/decorators/tokenPaymaster/getTokenPaymasterQuotes.ts:48]: BicoTokenPaymasterConfigParams

**Review by joepegler — CHANGES_REQUESTED** [review]
Some minor tweaks, but looking good now. Nice job 👍

File: src/sdk/account/utils/Utils.ts
Comment (joepegler) [inline src/sdk/account/utils/Utils.ts:423]: remove

File: src/sdk/account/utils/Utils.ts
Comment (joepegler) [inline src/sdk/account/utils/Utils.ts:419]: hmmm, doesn't BundlerClient already have an optional paymasterContext? You can remove this I think?

File: src/sdk/account/utils/Utils.ts
Comment (joepegler) [inline src/sdk/account/utils/Utils.ts:415]: Can remove?

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (joepegler) [inline src/sdk/clients/createBicoPaymasterClient.test.ts:92]: Do you need to cast this? Just curious because if you do we should fix 🤔

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (joepegler) [inline src/sdk/clients/createBicoPaymasterClient.test.ts:270]: To me this reads like the quote is not used and that the amount might ultimately be a bit different. It reads like the quote is acquired, and then discarded

**Review by joepegler — CHANGES_REQUESTED** [review]

**Comment by joepegler** [comment]
Hmmm, still some things I'm not sure about... Please check the thing about the quote being acquired and then appearing to be discarded. Am I reading it wrong?

**Comment by joepegler** [comment]
review (iv)

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (joepegler) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:270]: This could cause sporadic approval failures if this is doing what I think it is doing.

File: src/sdk/clients/createBicoPaymasterClient.test.ts
Comment (VGabriel45) [reply src/sdk/clients/createBicoPaymasterClient.test.ts:92]: No, fixed

**Pull request merged**
